File: src/buildenv.rs
use std::path::Path;
use std::str::FromStr;
use std::fs;

use serde::{Serialize, Deserialize};
use structopt::StructOpt;


use crate::
{
    core,
    cmake,
    printer
};


// list of compilers
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub enum Compiler
{
    VisualStudio2015,
    VisualStudio2017,
    VisualStudio2019,
    VisualStudio2022
}


// list of platforms
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub enum Platform
{
    Auto,
    Win32,
    X64
}


impl FromStr for Compiler {

    type Err = String;

    fn from_str(input: &str) -> Result<Compiler, Self::Err> {
        match input.to_ascii_lowercase().as_str() {
            "vs2015" => Ok(Compiler::VisualStudio2015),
            "vs2017" => Ok(Compiler::VisualStudio2017),
            "vs2019" => Ok(Compiler::VisualStudio2019),
            "vs2022" => Ok(Compiler::VisualStudio2022),
            // github actions installed compiler
            "windows-2016" => Ok(Compiler::VisualStudio2017),
            "windows-2019" => Ok(Compiler::VisualStudio2019),
            _        => Err("invalid compiler".to_string()),
        }
    }
}


impl FromStr for Platform {

    type Err = String;

    fn from_str(input: &str) -> Result<Platform, Self::Err> {
        match input.to_ascii_lowercase().as_str() {
            "auto"  => Ok(Platform::Auto),
            "win32" => Ok(Platform::Win32),
            "x64"   => Ok(Platform::X64),
            "win64" => Ok(Platform::X64),
            _       => Err("Invalid platform".to_string()),
        }
    }
}


fn is_64bit(platform: &Platform) -> bool
{
    match platform
    {
        Platform::Auto => {core::is_64bit()},
        Platform::Win32 => {false},
        Platform::X64 => {true}
    }
}


fn create_cmake_arch(platform: &Platform) -> &'static str
{
    if is_64bit(platform)
    {
        "x64"
    }
    else
    {
        "Win32"
    }
}

// gets the visual studio cmake generator argument for the compiler and platform
fn create_cmake_generator(compiler: &Compiler, platform: &Platform) -> cmake::Generator
{
    match compiler
    {
        Compiler::VisualStudio2015 =>
        {
            if is_64bit(platform) { cmake::Generator::new("Visual Studio 14 2015 Win64") }
            else { cmake::Generator::new("Visual Studio 14 2015") }
        },
        Compiler::VisualStudio2017 =>
        {
            if is_64bit(platform) { cmake::Generator::new("Visual Studio 15 Win64") }
            else { cmake::Generator::new("Visual Studio 15") }
        },
        Compiler::VisualStudio2019 => { cmake::Generator::new_with_arch("Visual Studio 16 2019", create_cmake_arch(platform))},
        Compiler::VisualStudio2022 => { cmake::Generator::new_with_arch("Visual Studio 17 2022", create_cmake_arch(platform))}
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct BuildEnviroment
{
    pub compiler: Option<Compiler>,
    pub platform: Option<Platform>
}

#[derive(StructOpt, Debug)]
pub struct EnviromentArgument
{
    #[structopt(long)]
    pub compiler: Option<Compiler>,

    #[structopt(long)]
    pub platform: Option<Platform>,

    #[structopt(long)]
    pub force: bool
}

impl BuildEnviroment
{
    pub fn new_empty() -> BuildEnviroment
    {
        BuildEnviroment
        {
            compiler: None,
            platform: None
        }
    }

    pub fn get_cmake_generator(&self) -> cmake::Generator
    {
        create_cmake_generator(self.compiler.as_ref().unwrap(), self.platform.as_ref().unwrap())
    }

    // update the build environment from an argparse namespace
    pub fn update_from_args(&mut self, printer: &mut printer::Printer, args: &EnviromentArgument)
    {
        match &args.compiler
        {
            None => {},
            Some(new_compiler) =>
            {
                match &self.compiler
                {
                    None =>  {self.compiler = Some(new_compiler.clone());},
                    Some(current_compiler) =>
                    {
                        if current_compiler != new_compiler
                        {
                            if args.force
                            {
                                printer.warning(format!("Compiler changed via argument from {:#?} to {:#?}", current_compiler, new_compiler).as_str());
                                self.compiler = Some(new_compiler.clone());
                            }
                            else
                            {
                                printer.error(format!("Compiler changed via argument from {:#?} to {:#?}", current_compiler, new_compiler).as_str());
                            }
                        }
                    }
                }
            }
        }

        match &args.platform
        {
            None => {},
            Some(new_platform) =>
            {
                match &self.platform
                {
                    None =>  {self.platform = Some(new_platform.clone());},
                    Some(current_platform) =>
                    {
                        if current_platform != new_platform
                        {
                            if args.force
                            {
                                printer.warning(format!("Platform changed via argument from {:#?} to {:#?}", current_platform, new_platform).as_str());
                                self.platform = Some(new_platform.clone());
                            }
                            else
                            {
                                printer.error(format!("Platform changed via argument from {:#?} to {:#?}", current_platform, new_platform).as_str());
                            }
                        }
                    }
                }
            }
        }
    }


    // validate the build environment
    pub fn validate(&self, printer: &mut printer::Printer) -> bool
    {
        let mut status = true;
        
        if self.compiler == None
        {
            printer.error("Compiler not set");
            status = false;
        };
        
        if self.platform == None
        {
            printer.error("Platform not set");
            status = false;
        };

        status
    }

    pub fn save_to_file(&self, path: &Path)
    {
        let data = serde_json::to_string(&self).unwrap();
        fs::write(path, data).expect("Unable to write file");
    }
}

// load build enviroment from json file
pub fn load_from_file(path: &Path, printer: Option<&mut printer::Printer>) -> BuildEnviroment
{
    match core::read_file_to_string(path)
    {
        Some(content) =>
        {
            let data : Result<BuildEnviroment, serde_json::error::Error> = serde_json::from_str(&content);
            match data
            {
                Ok(loaded) => loaded,
                Err(error) =>
                {
                    if let Some(pr) = printer
                    {
                        pr.error(format!("Unable to parse json {}: {}", path.to_string_lossy(), error).as_str());
                    }
                    BuildEnviroment::new_empty()
                }
            }
        },
        None => BuildEnviroment::new_empty()
    }
}

File: src/BuildEnv.cs
using Spectre.Console.Cli;
using System.ComponentModel;

namespace Workbench;

[TypeConverter(typeof(CompilerConverter))]
public enum Compiler
{
    VisualStudio2015,
    VisualStudio2017,
    VisualStudio2019,
    VisualStudio2022
}

class CompilerConverter : EnumTypeConverter<Compiler>
{
    public CompilerConverter()
    {
        // fallbacks: github actions installed compiler
        Data
            .Add(Compiler.VisualStudio2015, "vs2015")
            .Add(Compiler.VisualStudio2017, "vs2017", "windows-2016")
            .Add(Compiler.VisualStudio2019, "vs2019", "windows-2019")
            .Add(Compiler.VisualStudio2022, "vs2022")
            ;
    }
}


[TypeConverter(typeof(PlatformConverter))]
public enum Platform
{
    Auto,
    Win32,
    X64
}

class PlatformConverter : EnumTypeConverter<Platform>
{
    public PlatformConverter()
    {
        Data
            .Add(Platform.Auto, "auto")
            .Add(Platform.Win32, "win32", "x86")
            .Add(Platform.X64, "win64", "x64")
            ;
    }
}

// #[derive(Serialize, Deserialize, Debug)]
public class BuildEnviroment
{
    public Compiler? compiler = null;
    public Platform? platform = null;

    public static BuildEnviroment CreateEmpty()
    {
        return new BuildEnviroment();
    }

    public CMake.Generator CreateCmakeGenerator()
    {
        if (compiler == null) { throw new ArgumentNullException(nameof(compiler)); }
        if (platform == null) { throw new ArgumentNullException(nameof(platform)); }
        return BuildUitls.CreateCmakeGenerator(compiler.Value, platform.Value);
    }

    // validate the build environment
    public bool Validate(Printer printer)
    {
        var status = true;

        if (compiler == null)
        {
            printer.error("Compiler not set");
            status = false;
        }

        if (platform == null)
        {
            printer.error("Platform not set");
            status = false;
        }

        return status;
    }

    // update the build environment from an argparse namespace
    public void UpdateFromArguments(Printer printer, EnviromentArgument args)
    {
        UpdateCompiler(printer, args);
        UpdatePlatform(printer, args);

        void UpdateCompiler(Printer printer, EnviromentArgument args)
        {
            if (args.Compiler == null) { return; }

            if (this.compiler == null)
            {
                this.compiler = args.Compiler;
                return;
            }

            if (args.Compiler == this.compiler) { return; }

            if (args.ForceChange)
            {
                printer.warning($"Compiler changed via argument from {this.compiler} to {args.Compiler}");
                this.compiler = args.Compiler;
            }
            else
            {
                printer.error($"Compiler changed via argument from {this.compiler} to {args.Compiler}");
            }
        }

        void UpdatePlatform(Printer printer, EnviromentArgument args)
        {
            if (args.Platform == null) { return; }

            if (this.platform == null)
            {
                this.platform = args.Platform;
                return;
            }

            if (args.Platform == this.platform) { return; }

            if (args.ForceChange)
            {
                printer.warning($"Platform changed via argument from {this.platform} to {args.Platform}");
                this.platform = args.Platform;
            }
            else
            {
                printer.error($"Platform changed via argument from {this.platform} to {args.Platform}");
            }
        }
    }
}


public class EnviromentArgument : CommandSettings
{
    [Description("The compiler to use")]
    [CommandOption("--compiler")]
    [DefaultValue(null)]
    public Compiler? Compiler { get; set; }

    [Description("The platform to use")]
    [CommandOption("--platform")]
    [DefaultValue(null)]
    public Platform? Platform { get; set; }

    [Description("force a change if the compiler or platform differs from last time")]
    [CommandOption("--force")]
    [DefaultValue(false)]
    public bool ForceChange { get; set; }
}

public static class BuildUitls
{
    static bool Is64Bit(Platform platform)
    {
        return platform switch
        {
            Platform.Auto => Core.Is64Bit(),
            Platform.Win32 => false,
            Platform.X64 => true,
            _ => false,
        };
    }


    static string GetCmakeArchitctureArgument(Platform platform)
    {
        if (Is64Bit(platform))
        {
            return "x64";
        }
        else
        {
            return "Win32";
        }
    }

    // gets the visual studio cmake generator argument for the compiler and platform
    internal static CMake.Generator CreateCmakeGenerator(Compiler compiler, Platform platform)
    {
        switch (compiler)
        {
            case Compiler.VisualStudio2015:
                return Is64Bit(platform)
                    ? new CMake.Generator("Visual Studio 14 2015 Win64")
                    : new CMake.Generator("Visual Studio 14 2015")
                    ;
            case Compiler.VisualStudio2017:
                return Is64Bit(platform)
                    ? new CMake.Generator("Visual Studio 15 Win64")
                    : new CMake.Generator("Visual Studio 15")
                    ;
            case Compiler.VisualStudio2019:
                return new CMake.Generator("Visual Studio 16 2019", GetCmakeArchitctureArgument(platform));
            case Compiler.VisualStudio2022:
                return new CMake.Generator("Visual Studio 17 2022", GetCmakeArchitctureArgument(platform));
            default:
                throw new Exception("Invalid compiler");
        }
    }

    public static void SaveToFile(BuildEnviroment self, string path)
    {
        File.WriteAllText(path, JsonUtil.Write(self));
    }

    // load build enviroment from json file
    public static BuildEnviroment LoadFromFileOrCreateEmpty(string path, Printer printer)
    {
        if (File.Exists(path) == false)
        {
            return BuildEnviroment.CreateEmpty();
        }

        var content = File.ReadAllText(path);

        var loaded = JsonUtil.Parse<BuildEnviroment>(printer, path, content);
        if (loaded == null)
        {
            return BuildEnviroment.CreateEmpty();
        }

        return loaded;
    }
}
